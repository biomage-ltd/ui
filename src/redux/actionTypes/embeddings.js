const EMBEDDINGS = 'embeddings';

/**
 * Turns on the loading condition for the embedding.
 */
const EMBEDDINGS_LOADING = `${EMBEDDINGS}/loading`;

/**
 * Sets the state of the embedding to be successfully loaded, with the appropriate embedding data.
 */
const EMBEDDINGS_LOADED = `${EMBEDDINGS}/loaded`;

/**
 * Sets an error condition for a specific embedding.
 */
const EMBEDDINGS_ERROR = `${EMBEDDINGS}/error`;

export {
  EMBEDDINGS_LOADING, EMBEDDINGS_LOADED, EMBEDDINGS_ERROR,
};
