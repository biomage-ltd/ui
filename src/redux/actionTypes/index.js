const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC,
  SET_FOCUSED_GENE,
};
