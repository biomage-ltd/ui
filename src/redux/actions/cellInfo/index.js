import setCellInfoFocus from './setCellInfoFocus';
import updateCellInfo from './updateCellInfo';

export {
  setCellInfoFocus,
  updateCellInfo,
};
