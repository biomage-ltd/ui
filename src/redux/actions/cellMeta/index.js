import loadCellMeta from './loadCellMeta';

export default loadCellMeta;
