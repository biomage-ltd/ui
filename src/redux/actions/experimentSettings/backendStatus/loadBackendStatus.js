import fetchAPI from '../../../../utils/fetchAPI';
import endUserMessages from '../../../../utils/endUserMessages';
import { isServerError, throwIfRequestFailed } from '../../../../utils/fetchErrors';
import {
  EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADING,
  EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADED,
  EXPERIMENT_SETTINGS_BACKEND_STATUS_ERROR,

} from '../../../actionTypes/experimentSettings';

const loadBackendStatus = (experimentId) => async (dispatch) => {
  dispatch({
    type: EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADING,
    payload: {
      experimentId,
    },
  });

  const url = `/v1/experiments/${experimentId}/backendStatus`;
  try {
    const response = await fetchAPI(url);
    const status = await response.json();

    throwIfRequestFailed(response, status, endUserMessages.ERROR_FETCHING_STATUS);

    dispatch({
      type: EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADED,
      payload: {
        experimentId,
        status,
      },
    });

    return status;
  } catch (e) {
    if (!isServerError(e)) {
      console.error(`fetch ${url} error ${e.message}`);
    }
    dispatch({
      type: EXPERIMENT_SETTINGS_BACKEND_STATUS_ERROR,
      payload: {
        error: 'Could not get the status of the backend.',
        errorType: e,
      },
    });
  }
};

export default loadBackendStatus;
