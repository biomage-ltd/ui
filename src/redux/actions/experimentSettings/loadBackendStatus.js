import fetchAPI from '../../../utils/fetchAPI';
import {
  EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADING,
  EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADED,
  EXPERIMENT_SETTINGS_BACKEND_STATUS_ERROR,
} from '../../actionTypes/experimentSettings';

const loadBackendStatus = (experimentId) => async (dispatch) => {
  dispatch({
    type: EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADING,
    payload: {
      experimentId,
    },
  });

  try {
    const response = await fetchAPI(
      `/v1/experiments/${experimentId}/backendStatus`,
    );

    if (response.ok) {
      const status = await response.json();

      dispatch({
        type: EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADED,
        payload: {
          experimentId,
          status,
        },
      });

      return status;
    }

    throw new Error('HTTP status code was not 200.');
  } catch (e) {
    dispatch({
      type: EXPERIMENT_SETTINGS_BACKEND_STATUS_ERROR,
      payload: {
        error: 'Could not get the status of the backend.',
        errorType: e,
      },
    });
  }
};

export default loadBackendStatus;
