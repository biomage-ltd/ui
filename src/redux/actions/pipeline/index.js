import runPipeline from './runPipeline';
import runGem2s from './runGem2s';

export {
  runPipeline,
  runGem2s,
};
