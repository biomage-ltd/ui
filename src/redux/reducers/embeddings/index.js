import initialState from './initialState';
import {
  EMBEDDINGS_LOADED,
  EMBEDDINGS_LOADING,
  EMBEDDINGS_ERROR,
  EMBEDDINGS_RESET,
} from '../../actionTypes/embeddings';
import embeddingsLoading from './embeddingsLoading';
import embeddingsLoaded from './embeddingsLoaded';
import embeddingsError from './embeddingsError';
import embeddingsReset from './embeddingsReset';

const notificationsReducer = (state = initialState, action) => {
  switch (action.type) {
    case EMBEDDINGS_LOADING: {
      return embeddingsLoading(state, action);
    }

    case EMBEDDINGS_LOADED: {
      return embeddingsLoaded(state, action);
    }

    case EMBEDDINGS_ERROR: {
      return embeddingsError(state, action);
    }

    case EMBEDDINGS_RESET: {
      return embeddingsReset(state, action);
    }

    default: {
      return state;
    }
  }
};

export default notificationsReducer;
