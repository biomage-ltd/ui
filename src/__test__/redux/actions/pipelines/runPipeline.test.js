import _ from 'lodash';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import loadBackendStatus from '../../../../redux/actions/experimentSettings/backendStatus/loadBackendStatus';

import {
  EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADING,
  EXPERIMENT_SETTINGS_PIPELINE_START,
  EXPERIMENT_SETTINGS_BACKEND_STATUS_ERROR,
  EXPERIMENT_SETTINGS_DISCARD_CHANGED_QC_FILTERS,
} from '../../../../redux/actionTypes/experimentSettings';
import { EMBEDDINGS_LOADING } from '../../../../redux/actionTypes/embeddings';

import { runPipeline } from '../../../../redux/actions/pipeline';

import generateExperimentSettingsMock from '../../../test-utils/experimentSettings.mock';

const mockStore = configureStore([thunk]);

jest.mock('localforage');
enableFetchMocks();

jest.mock('../../../../redux/actions/experimentSettings/backendStatus/loadBackendStatus',
  () => jest.fn().mockImplementation(() => async () => { }));

const experimentId = 'experiment-id';

const sampleIds = ['sample1, sample2'];

const initialExperimentState = generateExperimentSettingsMock(sampleIds);

const initialState = {
  experimentSettings: {
    ...initialExperimentState,
    processing: {
      ...initialExperimentState.processing,
      meta: {
        changedQCFilters: new Set(['cellSizeDistribution']),
      },
    },
  },
};

describe('runPipeline action', () => {
  beforeEach(() => {
    const response = new Response(JSON.stringify({}));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Dispatches events properly', async () => {
    const store = mockStore(initialState);
    await store.dispatch(runPipeline(experimentId));

    const actions = store.getActions();

    expect(actions[0].type).toEqual(EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADING);
    expect(actions[1].type).toEqual(EXPERIMENT_SETTINGS_PIPELINE_START);
    expect(loadBackendStatus).toHaveBeenCalled();
    expect(actions).toMatchSnapshot();
  });

  it('Dispatches status error if loading fails', async () => {
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify({ message: 'some weird error that happened' }), { status: 400 });

    const store = mockStore(initialState);
    await store.dispatch(runPipeline(experimentId));

    const actions = store.getActions();

    expect(actions[0].type).toEqual(EXPERIMENT_SETTINGS_BACKEND_STATUS_LOADING);
    expect(loadBackendStatus).not.toHaveBeenCalled();
    expect(actions[1].type).toEqual(EXPERIMENT_SETTINGS_BACKEND_STATUS_ERROR);

    expect(actions).toMatchSnapshot();
  });

  it('Runs only the embedding if only changed filter was configureEmbedding', async () => {
    fetchMock.resetMocks();

    const onlyConfigureEmbeddingChangedState = _.cloneDeep(initialState);
    onlyConfigureEmbeddingChangedState.experimentSettings.processing.meta.changedQCFilters = new Set(['configureEmbedding']);

    const store = mockStore(onlyConfigureEmbeddingChangedState);
    await store.dispatch(runPipeline(experimentId));

    const actions = store.getActions();

    expect(actions[0].type).toEqual(EXPERIMENT_SETTINGS_DISCARD_CHANGED_QC_FILTERS);
    expect(actions[1].type).toEqual(EMBEDDINGS_LOADING);

    expect(actions).toMatchSnapshot();
  });

  // it("Clicking run with NO other filters changed doesn't trigger the pipeline", async () => {
  //   const store = mockStore(storeState);

  //   const component = mount(
  //     <Provider store={store}>
  //       <CalculationConfig
  //         experimentId='1234'
  //         width={50}
  //         height={50}
  //         onPipelineRun={mockOnPipelineRun}
  //         onConfigChange={mockOnConfigChange}
  //       />
  //     </Provider>,
  //   );
  //   act(() => { component.find(Select).at(0).getElement().props.onChange('tsne'); });
  //   component.update();

  //   const runButton = component.find(Button);
  //   runButton.simulate('click', {});
  //   expect(mockOnPipelineRun).toBeCalledTimes(0);
  //   await waitForActions(
  // store, [EXPERIMENT_SETTINGS_NON_SAMPLE_FILTER_UPDATE, EMBEDDINGS_LOADING]);
  //   expect(store.getActions()[1].type).toEqual(EMBEDDINGS_LOADING);
  // });
});
