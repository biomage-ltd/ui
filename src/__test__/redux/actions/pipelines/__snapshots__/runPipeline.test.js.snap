// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runPipeline action Dispatches events properly 1`] = `
Array [
  Object {
    "payload": Object {
      "experimentId": "experiment-id",
    },
    "type": "experimentSettings/backendStatusLoading",
  },
  Object {
    "payload": Object {},
    "type": "experimentSettings/startPipeline",
  },
]
`;

exports[`runPipeline action Dispatches status error if loading fails 1`] = `
Array [
  Object {
    "payload": Object {
      "experimentId": "experiment-id",
    },
    "type": "experimentSettings/backendStatusLoading",
  },
  Object {
    "payload": Object {
      "error": "Could not start the pipeline.",
      "errorType": "some weird error that happened",
    },
    "type": "experimentSettings/backendStatusError",
  },
]
`;

exports[`runPipeline action Runs only the embedding if only changed filter was configureEmbedding 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "experimentSettings/discardChangedQCFilters",
  },
  Object {
    "payload": Object {
      "embeddingType": "umap",
      "experimentId": "experiment-id",
    },
    "type": "embeddings/loading",
  },
]
`;
